### PR TITLE
Migrate to bison parser

### PR DESCRIPTION
**Pull request by LaPeste**
##  TODO

* [x] Changelog entry
* [ ] Tests (if applicable)

### CODE REVIEW

File: wrappers/src/results_cs.cpp
Comment (LaPeste) [inline wrappers/src/results_cs.cpp:137]: I had to brush up on move semantics. I don't think the `std::move` is needed in either of the case. One special rule says that if the move constructor and operator are defined, returning automatic l-values by copy from a method or function moves such values and not copy.
If my understanding is incorrect, then I'll put the `std::move` back where it was.

**Review by nirinchev — CHANGES_REQUESTED** [review]

File: wrappers/src/results_cs.cpp
Comment (nirinchev) [inline wrappers/src/results_cs.cpp:131]: [nit] Since `table` is used only once, we can remove the extra variable here and use `results.get_table().query(...)` on line 133.

**Review by nirinchev — APPROVED** [review]

**Comment by nirinchev** [comment]
Needs a changelog entry and I asked @jedelbo to double check the C++ code in case we're missing something about the new API.

File: wrappers/src/results_cs.cpp
Comment (jedelbo) [inline wrappers/src/results_cs.cpp:135]: `ordering` can be NULL!

**Review by jedelbo — CHANGES_REQUESTED** [review]

File: Tests/Realm.Tests/Database/CollectionTests.cs
Comment (nirinchev) [inline Tests/Realm.Tests/Database/CollectionTests.cs:614]: This is a regression - the change in wording is fine, but we shouldn't be surfacing the internal naming of tables. Might need to ping @ironage or @jedelbo about the best way to convert these error messages to use the expected type names.

File: CHANGELOG.md
Comment (nirinchev) [inline CHANGELOG.md]: This should also have a user-facing note. Perhaps sync with @jedelbo on what would be an appropriate user-facing note about it. As far as I can see, the package is now 6% smaller, so perhaps something like:

> Replaced the implementation of the string query parser (the one used for [realm.All<Foo>().Filter("some-string-query")](https://docs.mongodb.com/realm-sdks/dotnet/10.0.0-beta.3/reference/Realms.CollectionExtensions.html#Realms_CollectionExtensions_Filter__1_System_Linq_IQueryable___0__System_String_)). This results in ~5% reduction of the size of the native binary while keeping the query execution times on par with the old parser. (**if Jørgen has other comments here, be sure to include them**).

File: wrappers/src/results_cs.cpp
Comment (nirinchev) [inline wrappers/src/results_cs.cpp:135]: If ordering is `null`, you should probably fall back to the existing ordering defined on the results: `DescriptorOrdering ordering = query.get_ordering() ? *query.get_ordering() :results. get_descriptor_ordering();` (not sure if C++ will like this syntax, but I hope the idea is clear).

File: CHANGELOG.md
Comment (jedelbo) [reply CHANGELOG.md]: This is just fine

**Review by jedelbo — APPROVED** [review]

File: Tests/Realm.Tests/Database/CollectionTests.cs
Comment (ironage) [reply Tests/Realm.Tests/Database/CollectionTests.cs:614]: we missed a few places, it should be fixed by https://github.com/realm/realm-core/pull/4290

File: wrappers/src/results_cs.cpp
Comment (ironage) [inline wrappers/src/results_cs.cpp:133]: query.get_ordering() returns the ordering, but it also removes it from the query, so you can't call it twice (this is a bit of an unusual pattern). Try this: 
```
    query_parser::KeyPathMapping mapping;
    realm::populate_keypath_mapping(mapping, *this->get_realm());
    Query parsed_query = get_query().get_table()->query(query_string, no_args, mapping);
    DescriptorOrdering new_order = results.get_descriptor_ordering();
    if (auto parsed_ordering = parsed_query.get_ordering()) {
        new_order.append(*parsed_ordering);
    }
    return Results(m_realm, results.get_query().and_query(std::move(parsed_query)), new_order);
```
I'm adding the append in https://github.com/realm/realm-core/pull/4291

File: Tests/Realm.Tests/Database/CollectionTests.cs
Comment (LaPeste) [reply Tests/Realm.Tests/Database/CollectionTests.cs:614]: Thank you. Then I'll update core in our SDK to point to the right commit when the PR 4290 is merged. And of course change the string in the unit test.

File: wrappers/src/results_cs.cpp
Comment (LaPeste) [reply wrappers/src/results_cs.cpp:133]: Thank you very much. I've modified as you suggested and I'll wait for the PR 4291 to be merged, to try test the code.

Comment (nirinchev) [reply wrappers/src/results_cs.cpp:133]: You can temporarily point this PR to James's branch to get things building.

Comment (LaPeste) [reply wrappers/src/results_cs.cpp:133]: As expected it works. Thank you.

File: wrappers/src/results_cs.cpp
Comment (nirinchev) [inline wrappers/src/results_cs.cpp:138]: `std::move(new_order)`?

**Review by nirinchev — APPROVED** [review]

**Review by nirinchev — CHANGES_REQUESTED** [review]
Oops, hit approve too fast - still needs a user-facing changelog entry.

File: wrappers/src/results_cs.cpp
Comment (LaPeste) [reply wrappers/src/results_cs.cpp:138]: I don't know, to be honest. I noticed that as well but @irongage didn't use it in the example. I thought it was a stylistic error, but then I went to check and only `Query` has a custom move constructor, while `DescriptorOrdering` has only the default one. After seeing that I thought that it was not a mistake but a wanted choice.
What is your opinion?

File: CHANGELOG.md
Comment (nirinchev) [inline CHANGELOG.md:11]:
```suggestion
* Replaced the implementation of the string query parser (the one used for [`realm.All().Filter("some-string-query")`](https://docs.mongodb.com/realm-sdks/dotnet/10.0.0-beta.3/reference/Realms.CollectionExtensions.html#Realms_CollectionExtensions_Filter__1_System_Linq_IQueryable___0__System_String_)). This results in ~5% reduction of the size of the native binary while keeping the query execution times on par with the old parser. (PR [#2185](https://github.com/realm/realm-dotnet/pull/2185), Core upgrade)
```

At the end of the change entry we try to add some additional information, depending on the change type:
* If there's an issue filed for the change (typically a bug or a feature request), we add (Issue [#123](https://github.com/realm/realm-dotnet/issue/123)). This allows the person who filed the issue and everyone interested to see that the new release includes what they were asking for.
* If there's no issue we add a link to the PR that introduced the change - e.g. (PR [#2120](https://github.com/realm/realm-dotnet/pull/2120)) like above. This allows us to quickly find the context of a change by looking at the changelog, which is (typically) more readable than the git history.
* Finally, if all a PR does is to bump the core version and we get some fixes for free, we add (Core upgrade). 

In this case, I proposed a combination of 2 and 3 because the filesize reduction is thanks to the Core upgrade, but the PR includes some changes of its own, so worth linking to it.

As for the link to the API docs - we typically don't bother linking to the API from the changelog, but in this case, it concerns a feature that may not be super heavily used (as I'm fairly certain most devs will prefer to use LINQ), so adding a link makes it more prominent and will allow people who are curious about it to find more info quickly.

**Review by nirinchev — APPROVED** [review]
Final nit from me. Otherwise, should be ready to merge once https://github.com/realm/realm-core/pull/4290 goes in.

File: wrappers/src/results_cs.cpp
Comment (nirinchev) [reply wrappers/src/results_cs.cpp:138]: I don't know - I'm not a C++ dev 😄 As mentioned, I always add `std::move` when I want to move something, just to be on the safe side. Maybe @ironage can educate us on the pros/cons of `std::move`-ing the ordering?

Comment (LaPeste) [reply wrappers/src/results_cs.cpp:138]: That feedback would be very welcome. 🙂

Comment (ironage) [reply wrappers/src/results_cs.cpp:138]: Yes you can move the ordering, I just forgot. Either way is functionally correct, but moving it is an efficient transfer of ownership, while not moving it invokes the [copy constructor ](https://github.com/realm/realm-core/blob/master/src/realm/sort_descriptor.cpp#L545-L550) which inefficiently makes copies of all the contents into the new object only to then discard the previous object entirely.

Comment (LaPeste) [reply wrappers/src/results_cs.cpp:138]: My only doubt was coming from the fact that while brushing on `std::move` I read that returning automatic local variables generally (not sure if always) uses the move semantics instead of the copy ones, as long as a move constructor and assignment exist. So I was not sure if stylistically people still added a `std::move` on the returned values. You clearly answered that people generally prefer to be explicit, thank you for the feedback.

**Pull request merged**
